text/108-bullet-indent.md

```diff
@@ -0,0 +1,260 @@
+Proof.
+split.
+  exact I.
+split.
+  exact I.
+split.
+  exact I.
+split.
+- exact I.
+- exact I.
+Qed.
+```
+
+This scheme happens a lot in actual mathematical proofs with a few
+simple side conditions, handled first, then a -- more difficult --
+main goal, constituting the bulk of the proof.
+
+# Detailed design
+
+We add a new bullet behavior:
+
+```Coq
+Set Bullet Behavior "Indent".
+```
+
+## Example
+
+```Coq
+Set Default Goal Selector "!".
+Set Bullet Behavior "Indent".
+
+Goal True /\ True /\ (True /\ True).
+Proof.
+split.
+  exact I.  (* positive indentation, exactly two goals, focusing first *)
+split; [|split].  (* last subgoal, using negative indentation *)
+- exact I. (* regular use of bullets *)
+- exact I.
+exact I.  (* last subgoal, using negative indentation *)
+Qed.
+```
+
+## Main ideas
+
+In addition to current bullets:
+* a *positive indentation* (a line being more indented than previous one)
+  acts as a bullet when there are exactly two focused goals ;
+* a *negative indentation* (a line being less indented than previous one)
+  acts as a last bullet (when there is exactly one focused goal)
+
+## Algorithm
+
+We use a stack of
+
+```ocaml
+type indent =
+  | IndentBullet of bullet_type * int
+  | IndentTactic of int
+```
+
+When encountering a new bullet `b` at indentation `indent` we
+essentially handle it as of today, with a few additional warnings:
+* look in the stack for the same bullet `IdentBullet(b, pre_indent)`
+  * when found, warn if `indent != pre_indent`
+  * otherwise, if the top of the stack is `IndentTactic pre_indent`,
+    warn if `indent != pre_indent`
+* then, warn if a single goal is focused (the bullet is useless)
+* then proceed with current algorithm for bullets.
+
+When encountering a tactic at indentation `indent`:
+* if the top of the stack is `IndentBullet (_, pre_indent)`,
+  warn if `indent <= pre_indent`, focus the first goal
+  and add `IndentTactic indent` of top of the stack
+* otherwise, the top of the stack is `IndentTactic pre_indent`
+  * if `indent = pre_indent`, do nothing
+  * if `indent > pre_indent`, we have a *positive indentation*,
+    * raise an error if there are not exactly two focused goals
+    * then, focus the first goal and add `IndentTactic indent` of top of the stack
+  * if `indent < pre_indent`, we have a *negative indentation*,
+    * raise an error if there is any focused goal
+    * then, repeatedly pop the stack until we get a focused goal on top of it
+      (raise an error if it becomes empty)
+    * warn if `indent` doesn't match the indentation on top of the stack
+    * raise an error if there isn't exactly one focused goal
+    * then, focus the first goal and add `IndentTactic indent` of top of the stack.
+
+### Implementation details / notes
+
+* The implementation will mostly happen in `proof_bullet.ml`, with
+  some additional tracking of indentation in higher levels that can
+  likely reuse current locations.
```

Comment:
Glad you submitted this RFC, it is long overdue !

This part may not be so easy to implement tho, given the status of the higher-layers.

A possibility that would help inform implementation choices betters is to write a linter first.

I was thinking about how to do this in `coq-lsp` and I think I got a clear design. Let me know if you are interested and we can give it a go.